### DISCUSSION TITLE
get_oauth_token.php does not result in refresh token... redirects to homepage... no error

### DISCUSSION THREAD

**Discussion by mohsincheema** [PHPMailer/PHPMailer#2645] [General]
Hi,
I am trying to use PHPMailer get_oauth_token.php to enable XOAUTH2 authorization for gmail in Open Journal System (OJS) software published by PKP. As guided here at https://github.com/PHPMailer/PHPMailer/wiki/Using-Gmail-with-XOAUTH2 , I followed all the steps and when I run the vendor/phpmailer/phpmailer/get_oauth_token.php in browser and complete the subsequent steps fo selecting provider (google) then sign in, then allow user consent, it results in redirection to homepage and does not display the refresh token. No error is showing in browser or error log. I am unable to figure out what is going wrong. I have ask the OJS PKP support, and they said I should ask PHPMailer support. 
Please help me to resolve this issue. Thanks in anticipation.

**Comment by mohsincheema** [comment]
Code to reproduce is:

`<?php

/**

PHPMailer - PHP email creation and transport class.
PHP Version 5.5
@Package PHPMailer
@see https://github.com/PHPMailer/PHPMailer/ The PHPMailer GitHub project
https://github.com/author Marcus Bointon (Synchro/coolbru) [phpmailer@synchromedia.co.uk](mailto:phpmailer@synchromedia.co.uk)
https://github.com/author Jim Jagielski (jimjag) [jimjag@gmail.com](mailto:jimjag@gmail.com)
https://github.com/author Andy Prevost (codeworxtech) [codeworxtech@users.sourceforge.net](mailto:codeworxtech@users.sourceforge.net)
https://github.com/author Brent R. Matzelle (original founder)
@copyright 2012 - 2020 Marcus Bointon
@copyright 2010 - 2012 Jim Jagielski
@copyright 2004 - 2009 Andy Prevost
@license http://www.gnu.org/copyleft/lesser.html GNU Lesser General Public License
@note This program is distributed in the hope that it will be useful - WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.
*/
/**

Get an OAuth2 token from an OAuth2 provider.
Install this script on your server so that it's accessible
as [https/http]:////get_oauth_token.php
e.g.: http://localhost/phpmailer/get_oauth_token.php
Ensure dependencies are installed with 'composer install'
Set up an app in your Google/Yahoo/Microsoft account
Set the script address as the app's redirect URL
If no refresh token is obtained when running this file,
revoke access to your app and run the script again.
*/
namespace PHPMailer\PHPMailer;

/**

Aliases for League Provider Classes
Make sure you have added these to your composer.json and run composer install
Plenty to choose from here:
@see http://oauth2-client.thephpleague.com/providers/thirdparty/
*/
// @see https://github.com/thephpleague/oauth2-google
use League\OAuth2\Client\Provider\Google;
// @see https://packagist.org/packages/hayageek/oauth2-yahoo
use Hayageek\OAuth2\Client\Provider\Yahoo;
// @see https://github.com/stevenmaguire/oauth2-microsoft
use Stevenmaguire\OAuth2\Client\Provider\Microsoft;
if (!isset($_GET['code']) && !isset($_GET['provider'])) {
?>

Select Provider:
[Google](https://github.com/PHPMailer/PHPMailer/issues/2646?provider=Google)
[Yahoo](https://github.com/PHPMailer/PHPMailer/issues/2646?provider=Yahoo)
[Microsoft/Outlook/Hotmail/Live/Office365](https://github.com/PHPMailer/PHPMailer/issues/2646?provider=Microsoft)
//require 'vendor/autoload.php';
require '/home/username/public_html/subdomain/lib/pkp/lib/vendor/autoload.php';
session_start();

$providerName = '';

if (array_key_exists('provider', $_GET)) {
$providerName = $_GET['provider'];
$_SESSION['provider'] = $providerName;
} elseif (array_key_exists('provider', $_SESSION)) {
$providerName = $_SESSION['provider'];
}
if (!in_array($providerName, ['Google', 'Microsoft', 'Yahoo'])) {
exit('Only Google, Microsoft and Yahoo OAuth2 providers are currently supported in this script.');
}

//These details are obtained by setting up an app in the Google developer console,
//or whichever provider you're using.
$clientId = 'entered the clientID here';
$clientSecret = 'entered the clientSecret here';

//If this automatic URL doesn't work, set it yourself manually to the URL of this script
//$redirectUri = (isset($_SERVER['HTTPS']) ? 'https://' : 'http://') . $_SERVER['HTTP_HOST'] . $_SERVER['PHP_SELF'];
$redirectUri = 'https://sd.mydomainname.com/lib/pkp/lib/vendor/phpmailer/phpmailer/get_oauth_token.php';

$params = [
'clientId' => $clientId,
'clientSecret' => $clientSecret,
'redirectUri' => $redirectUri,
'accessType' => 'offline'
];

$options = [];
$provider = null;

switch ($providerName) {
case 'Google':
$provider = new Google($params);
$options = [
'scope' => [
'https://mail.google.com/'
]
];
break;
case 'Yahoo':
$provider = new Yahoo($params);
break;
case 'Microsoft':
$provider = new Microsoft($params);
$options = [
'scope' => [
'wl.imap',
'wl.offline_access'
]
];
break;
}

if (null === $provider) {
exit('Provider missing');
}

if (!isset($_GET['code'])) {
// If we don't have an authorization code then get one
$authUrl = $provider->getAuthorizationUrl($options);
$_SESSION['oauth2state'] = $provider->getState();
header('Location: ' . $authUrl);
exit;
// Check given state against previously stored one to mitigate CSRF attack
} elseif (empty($_GET['state']) || ($_GET['state'] !== $_SESSION['oauth2state'])) {
unset($_SESSION['oauth2state']);
unset($_SESSION['provider']);
exit('Invalid state');
} else {
unset($_SESSION['provider']);
// Try to get an access token (using the authorization code grant)
$token = $provider->getAccessToken(
'authorization_code',
[
'code' => $_GET['code']
]
);
// Use this to interact with an API on the users behalf
// Use this to get a new access token if the old one expires
echo 'Refresh Token: ', $token->getRefreshToken();`

**Comment by decomplexity** [comment]
Worth checking first whether the $redirectUri you specified in your script (above) is EXACTLY the same as the one specified to the Google console.

**Reply by Synchro** [reply]
What @decomplexity said. Also the redirect only needs to be accessible from *your* machine (Google will just send you a redirect to go there, it doesn't have to go there itself), so it's actually possible to use `localhost` URLs. Whatever URL you use for this script requesting the token should be the same as the redirect URL.

**Reply by mohsincheema** [reply]
> reply

Thanks decomplexity! I have checked again the $redirectUri in get_oauth_token.php is exactly the same as Authorized redirect URI under Client ID for Web application in Google APIs and Services. Also if both these don't match, it usually returns "Authorization Error Error 400: redirect_uri_mismatch" when we select the provider while running get_oauth_token.php in browser.

**Reply by mohsincheema** [reply]
> What @decomplexity said. Also the redirect only needs to be accessible from _your_ machine (Google will just send you a redirect to go there, it doesn't have to go there itself), so it's actually possible to use `localhost` URLs. Whatever URL you use for this script requesting the token should be the same as the redirect URL.

dear @Synchro I had tested this whole thing in localhost where I installed the OJS in a folder and run the whole process (keeping the get_oauth_token.php in its original place ./lib/pkp/lib/phpmailer/phpmailer/get_oauth_token.php) and the process behaved exactly same except the last: in localhost environment it displayed the refresh token (instead of redirecting to homepage) after allow user permission screen. but in live website it redirects to homepage. I don't know where I am making a mistake.

**Reply by Synchro** [reply]
I don't know where your issue is, but you probably don't need to make it work on the live site if you can make it work on localhost. Once you have the access and refresh tokens you can install them on your live server where they will work fine. They are not tied to your callback URL.

**Reply by mohsincheema** [reply]
@Synchro thanks, i shall test this workaround but would still love if we can fix the issue as well. I shall update you on it. previously I had tested this on localhost using a free gmail account. but on live site we are creating client id using our google workspace account.

**Reply by decomplexity** [reply]
To check: in the live site you are using a different _clientId_ and _clientSecret_ and well as different _redirectUri_? And it is only the refresh token that matters to the OAuth2 code since since it is used on each code invocation to get a new access token.

**Reply by mohsincheema** [reply]
> I don't know where your issue is, but you probably don't need to make it work on the live site if you can make it work on localhost. Once you have the access and refresh tokens you can install them on your live server where they will work fine. They are not tied to your callback URL.

@Synchro @decomplexity 

I have tried following: 
1. used the clientID and clientSecrect of our google workspace account (say, live.example@do_main.com), added a localhost/ojs/lib/pkp/lib/vendor/phpmailer/phpmailer/get_oauth_token.php in authorized uris in client ID for web application in google apis and services. 
2. used above mentioned (1.) clientID, clientSecret and redirect uri in get_oauth_token.php under localhost environment and run the get_oauth_token.php in browser under localhost environment
3. selected the provider, signed in with another email (google workspace account say, email2@do_main.com), allowed user permissions and resultantly, the refresh token was displayed.
4. copied this refresh token and clientId, clientSecret to config.inc.php of the OJS on live server
5. Then tried to send email from OJS workflow, it notified error sending email and following was recorded in error logs:

> [10-Mar-2022 17:00:36 UTC] PHP Warning:  Declaration of EmailReviewerForm::execute($submission) should be compatible with Form::execute(...$functionArgs) in /home/username/public_html/subdomain/lib/pkp/controllers/grid/users/reviewer/form/EmailReviewerForm.inc.php on line 18
[10-Mar-2022 17:00:52 UTC] PHP Warning:  Declaration of EmailReviewerForm::execute($submission) should be compatible with Form::execute(...$functionArgs) in /home/username/public_html/subdomain/lib/pkp/controllers/grid/users/reviewer/form/EmailReviewerForm.inc.php on line 18
[10-Mar-2022 17:00:53 UTC] SMTP connect() failed. https://github.com/PHPMailer/PHPMailer/wiki/Troubleshooting 

this thing is really delaying our progress....

**Reply by Synchro** [reply]
Those declaration errors have nothing to do with PHPMailer?

**Reply by mohsincheema** [reply]
seems so! i have raised the issue with OJS support as well. looking forward to some solution soon!
